### PR TITLE
Elegibility & Expiration as u64 API change

### PR DESCRIPTION
**Pull request by CPerezz**
Review & merge once #89 has been merged already as this PR relies on it.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,20 +1,46 @@
...
+dusk-jubjub = {version = "0.5", default-features = false}
+dusk-plonk = {version = "0.3", features = ["trace-print"], optional = true}
+num-bigint = {version = "0.3", optional = true } 
+num-traits = {version = "0.2", optional = true } 
+plonk_gadgets = {git = "https://github.com/dusk-network/plonk_gadgets", tag = "v0.4.3"}
+rand_core = {version = "0.5", default-features = false}  
 lazy_static = "1.4.0"
-rand = "0.7"
-nstack = "0.5"
+rand = {version = "0.7", default-features = false }
```
Comment (vlopes11) [inline Cargo.toml:18]: `rand` it not `no-std` compatible and should be optional

File: Cargo.toml [hunk 2]
```diff
@@ -1,20 +1,46 @@
...
+canonical = { version = "0.4", optional = true }
+canonical_derive = { version = "0.4", optional = true }
+anyhow = {version = "1", optional = true}
```
Comment (vlopes11) [inline Cargo.toml:21]: `anyhow` should be avoided as already discussed due to the big amount of involved problems already discussed

File: src/bid/encoding.rs [hunk 3]
```diff
@@ -9,58 +9,66 @@
 //! See: https://hackmd.io/@7dpNYqjKQGeYC7wMlPxHtQ/BkfS78Y9L
 
 use super::Bid;
+use dusk_bls12_381::BlsScalar;
+#[cfg(feature = "std")]
 use dusk_plonk::constraint_system::ecc::Point as PlonkPoint;
+#[cfg(feature = "std")]
 use dusk_plonk::prelude::*;
-use poseidon252::{sponge::sponge::*, StorageScalar};
+use poseidon252::sponge::sponge::*;
```
Comment (vlopes11) [inline src/bid/encoding.rs:17]: `poseidon252::sponge::sponge::sponge_hash` is deprecated, as discussed. `poseidon252::sponge::hash` should be used instead

File: src/bid/encoding.rs [hunk 4]
```diff
@@ -9,58 +9,66 @@
...
-const TYPE_FIELDS: [u8; 32] = *b"53313110000000000000000000000000";
+const TYPE_FIELDS: [u8; 32] = *b"53313116000000000000000000000000";
 
 impl Bid {
     /// Calculate the one-way BlsScalar representation of the Bid
     pub fn hash(&self) -> BlsScalar {
         // Generate an empty vector of `Scalar` which will store the
         // representation of all of the `Bid` elements.
-        let mut words_deposit = Vec::new();
+        let mut words_deposit = [BlsScalar::zero(); 13];
```
Comment (vlopes11) [inline src/bid/encoding.rs:36]: We should have two functions: `hash_inputs` that will output the hash inputs of the `Bid` structure, suck as `fn hash_inputs(&self) -> [BlsScalar; 13]`. The second one is the `hash` that will `fn hash(&self) -> BlsScalar` and use internally the `hash_inputs`

This is due to the fact that we may need to perform the hash outside the library with the `PoseidonLeaf` trait contract implementation

File: src/errors.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    MinimumBidValueUnreached {
+        /// The minimum bid_value required
+        min_val: JubJubScalar,
+        /// The expected length
+        found: JubJubScalar,
+    },
+    /// Error when there is a decrypt attempt with the wrong secret
+    WrongSecretProvided,
+    /// Invalid encoding/decoding
+    IOError,
```
Comment (vlopes11) [inline src/errors.rs:41]: Do we still have I/O anywhere?

File: src/proof.rs [hunk 6]
```diff
@@ -5,21 +5,21 @@
...
 };
-use dusk_plonk::jubjub::{
-    AffinePoint, GENERATOR_EXTENDED, GENERATOR_NUMS_EXTENDED,
-};
 use dusk_plonk::prelude::*;
 use plonk_gadgets::{AllocatedScalar, RangeGadgets::max_bound};
 use poseidon252::{
-    merkle_proof::merkle_opening_gadget, sponge::sponge::*, PoseidonBranch,
-    StorageScalar,
+    sponge::sponge::*,
```
Comment (vlopes11) [inline src/proof.rs:21]: Deprecated usage, check above

File: src/score_gen/score.rs [hunk 7]
```diff
@@ -6,20 +6,30 @@
...
 //! Score generation
 
-use super::errors::ScoreError;
 use super::{MINUS_ONE_MOD_2_POW_128, SCALAR_FIELD_ORD_DIV_2_POW_128};
 use crate::bid::Bid;
-use anyhow::{Error, Result};
-use dusk_plonk::jubjub::AffinePoint;
+use crate::errors::BlindBidError;
+#[cfg(feature = "std")]
+use anyhow::{Error as AnyhowError, Result};
```
Comment (vlopes11) [inline src/score_gen/score.rs:13]: Avoid anyhow, check above

**Review by vlopes11 — CHANGES_REQUESTED** [review]
Few changes required, mostly to be compatible with the `no-std` environment

File: Cargo.toml (hunk 2 above)
Comment (CPerezz) [reply Cargo.toml:21]: I forgot the `optional`. Anyway, it is needed in the `std` side to deal with the error handling for the circuits.

File: src/errors.rs (hunk 5 above)
Comment (CPerezz) [reply src/errors.rs:41]: It's for the to_bytes/from_bytes methods.

File: src/bid/encoding.rs (hunk 3 above)
Comment (CPerezz) [reply src/bid/encoding.rs:17]: I wanted to use the gadget. And there's no shorcut to the path needed. Anyway I refactored the imports in other places.

File: src/bid/encoding.rs (hunk 4 above)
Comment (CPerezz) [reply src/bid/encoding.rs:36]: Agree. We discussed that in the past I remember. Fixing it.

File: src/proof.rs (hunk 6 above)
Comment (CPerezz) [reply src/proof.rs:21]: Fixed.

File: Cargo.toml (hunk 1 above)
Comment (CPerezz) [reply Cargo.toml:18]: Done ty

**Review by vlopes11 — APPROVED** [review]
With the `fn hash_inputs(&self) -> [BlsScalar; 13]` fix, LGTM!

**Pull request merged**